### ISSUE TITLE
augur.js filters missing events

### ISSUE THREAD

**Issue by kevinday** [AugurProject/augur.js#135]
In looking into cache node sync issues, I'm noticing that not all market created events are getting passed along. To reproduce this locally, I set up my listeners:

```
var options = {
     http: "http://localhost:8545",
     ws: "ws://localhost:8546"
};

var augur = require("augur.js");

augur.connect(options);

augur.filters.listen({

    marketCreated: function (msg) {
        console.log("market created filter:", msg);
    },
    log_price: function (msg){
    	console.log("price filter:", msg);
    },
    tradingFeeUpdated: function (msg){
    	console.log("trading fee updated:", msg);
    },
    log_fill_tx: function (msg){
    	console.log("tx fill:", msg);
    }

}, function (filters) {
    console.log("setup complete:", filters);
});
```
@tinybike's market spammer is running, and initially I do see marketCreated events coming through. Eventually these stop (markets are still being created because getNumMarketsBranch is still incrementing). Not sure what is going on yet, but maybe the websocket connection is breaking at some point and events are no longer passed along?

I will run this same test except over RPC to see if I'm seeing the same behavior.

**Comment by tinybike** [comment]
Try using `log_price` instead of `price`, does that fix it?

**Comment by kevinday** [comment]
Oops, yeah my real code is actually using log_price, I just hadn't updated this example yet.

I think it's unrelated to this issue though because for this test I was only watching createMarket events, which were showing up at first but then eventually stopped coming through all together. Yeah, price is an invalid label but it just gets ignored on filter setup - it shouldn't cause createMarket events to eventually stop showing up.

**Comment by tinybike** [comment]
That does sound odd.  Questions:

- Are you using regular HTTP RPC or websockets?

When this stops occurring:

- Is the filter ID still set?  What value is it?  Can check using `augur.filters.filter.marketCreated`; it should have a non-null `id` field.
- If you're using HTTP RPC, is the `heartbeat` field (of the `filter.marketCreated` object) non-null?  If so, what is it?  It should be a `Timeout` object.  If it is, what are the values of `filter.marketCreated.heartbeat._called` and `filter.marketCreated.heartbeat._repeat.toString()`?
- If you're using websockets:
    - Each filter's message handler is registered in the `augur.rpc.subscriptions` object.  What is the value of `augur.rpc.subscriptions[augur.filters.filter.marketCreated.id].toString()`?
    - If you only use a single filter, the marketCreated filter, and set `augur.rpc.debug.broadcast = true`, are you still receiving `eth_subscribe` messages after the filter stops working?  If so, can you paste an example message here?

**Comment by tinybike** [comment]
Also, just so you know, my market spammer has been off for most of the day.  However I will turn it back on (and leave it on overnight).

**Comment by kevinday** [comment]
I'm using websockets when having the problems. Filters seem to work fine over RPC.

I just turned a marketCreated filter on in a debugger, so I should be able to answer the rest of your questions in a bit here.

**Comment by kevinday** [comment]
Ok, so I started filters with block and createMarket listeners. The `block` filter never stopped working, but the `createMarket` did. When `createMarket` stops working I see:

`augur.filters.filter.marketCreated.id` is `"0x173df39b2de0ee09d5d1f8fff3560c97"`
`augur.rpc.subscriptions[augur.filters.filter.marketCreated.id].toString()` is 
```
"function (msg) {
                                self.parse_marketCreated_message(msg, callback);
                            }"
```
With `augur.rpc.debug.broadcast = true` I do see the subscribe messages, but only when the filters are getting set up:
```
[ethrpc] WebSocket request to ws://localhost:8546 
{"id":4,"jsonrpc":"2.0","method":"eth_subscribe","params":["newBlocks",{"includeTransactions":false,"transactionDetails":false}]}
[ethrpc] WebSocket request to ws://localhost:8546 
{"id":5,"jsonrpc":"2.0","method":"eth_subscribe","params":["logs",{"address":"0x181ab5cfb79c3a4edd7b4556412b40453edeec32","topics":["0x63f140d7adcc464732c9379020aa9e5ce1b1e350796814d780ea3ca41d62a36b"]}]}

```
I think this makes sense though - I wouldn't to keep seeing ethrpc messages since I'm using websockets and events (should) be getting pushed, rather than polling via rpc.

So it appears that the `eth_subscribe` to the `logs` eventually stops working. `eth_subscribe` to `newBlocks` works fine.

For reference, my updated filter test code:
```
var options = {
     http: "http://localhost:8545",
     ws: "ws://localhost:8546"
};

var augur = require("augur.js");

augur.connect(options);
augur.rpc.debug.broadcast = true

augur.filters.listen({
	block: function (msg){
		console.log("block", msg);
	},
    marketCreated: function (msg) {
        if (msg) console.log("market created filter:", msg);
    }
}, function (filters) {
    console.log("setup complete:", filters);
});
```

**Comment by kevinday** [comment]
```
geth version
Geth
Version: 1.5.0-unstable-c8a8ad97
Protocol Versions: [63 62 61]
Network Id: 1
Go Version: go1.6.2
OS: darwin
GOPATH=
GOROOT=/usr/local/Cellar/go/1.6.2/libexec
```
for reference

**Comment by tinybike** [comment]
Sorry, the messages I meant to ask for were `eth_subscription` messages (not `eth_subscribe`, which as you point out, would not make sense).

**Comment by kevinday** [comment]
How do I see eth_subscription messages (I'm not seeing any).

**Comment by kevinday** [comment]
Will augur.js get notified if the websocket connection is disconnected? There was some discussion here about whether or not to enable auto-reconnect, but this issue was recently closed and I don't know what the resolution was: https://github.com/ethereum/EIPs/issues/51.

Looking at https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB#considerations it kinda sounds like the client is responsible for handling any unexpected disconnects (also, it really makes it sound like websockets aren't yet a great choice for cache node since cache nodes become useless if any events are missed at all).

**Comment by kevinday** [comment]
FYI, seeing this same behavior over IPC.

**Comment by kevinday** [comment]
Filed an issue with geth: https://github.com/ethereum/go-ethereum/issues/2887

**Comment by tinybike** [comment]
Temporary fix in cc2bb5ecefe8d27f09fb2234944f33f2e7f6c50a; will be permanently fixed when https://github.com/ethereum/go-ethereum/pull/2885 is merged.

**Comment by fjl** [comment]
It is merged to geth develop now. Please re-test.
